### DISCUSSION TITLE
Strange order of execution，There's no short circuit. What causes this？

### DISCUSSION THREAD

**Discussion by Arc-huangjingtong** [dotnet/csharplang#8288] [Q&A] [answered]
```Csharp
public class TestClass
{
    //change to int
    public static implicit operator int(TestClass c)
    {
        if (c == null)
        {
            throw new Exception();
        }

        return 1;
    }
}


public class Test
{
    public TestClass TestClass { get; set; }


    [Test]
    public void TestMethod()
    {
        int integer = false ? TestClass : default;

        Console.WriteLine(integer);//will throw Exception！
    }
}


```

**Comment by HaloFour** [comment]
With local type inference, the type of `integer` is `TestClass`, and the `default` value of `TestClass` is `null`.

```cs
var integer = false ? TestClass : default;
```

Did you mean?

```cs
int integer = false ? TestClass : default;
```

**Reply by Arc-huangjingtong** [reply]
thanks，but if you change “var”to “int” the exception is also throw

**Reply by HaloFour** [reply]
Yes, because you are trying to convert `default` of a `TestClass` to an `int`, and the conversion operator you have written will throw an exception in that case.  The conversion takes place after the conditional operator expression.

**Reply by TahirAhmadov** [reply]
@Arc-huangjingtong it's a different exception - the one you throw in your implicit operator; the `false ? TestClass : default` is evaluated to `null`, then `implicit operator` is called with `null` and you have:
```cs
        if (c == null)
        {
            throw new Exception();
        }
```

**Reply by TahirAhmadov** [reply]
You want: `int integer = false ? (int)TestClass : default;`
Or try this even: `int integer = false ? TestClass : 0;` (i'm not sure if this works, though)

**Reply by Arc-huangjingtong** [reply]
yes ，you change “defualt” to 0，will works

**Comment by Arc-huangjingtong** [comment]
i want to know why not circuit？
this is not intuitive

**Reply by HaloFour** [reply, marked as answer]
The expression does short-circuit.  That's not the problem.  The issue is that the expression by itself is evaluated before the assignment, which is when the conversion operation occurs.  The conditional expression `false ? TestClass : default` evaluates to a `TestClass`, not an `int`, and the `default` value of a `TestClass` is `null`.

**Reply by Arc-huangjingtong** [reply]
thanks， i see

**Reply by huoyaoyuan** [reply]
![image](https://github.com/user-attachments/assets/b30b05e4-411c-463c-9ebb-a4dd88385305)

The tooltip can inform you that the `default` resolves to `default(TestClass)`.
